### PR TITLE
Prohibit zero IDs

### PR DESCRIPTION
**Pull request by cthulhu-rider**

### CODE REVIEW

File: bearer/bearer_test.go [hunk 1]
```diff
@@ -220,11 +220,11 @@ func TestToken_AssertContainer(t *testing.T) {
 
 	eaclTable := eacltest.Table()
 
-	eaclTable.SetCID(cidtest.ID())
+	eaclTable.LimitToContainer(cidtest.ID())
```
Comment (smallhive) [inline bearer/bearer_test.go:223]: The previous name was better. It shows, you are setting the ID, not a complete container

File: client/container_statistic_test.go [hunk 2]
```diff
@@ -587,7 +587,7 @@ func TestClientStatistic_ObjectPut(t *testing.T) {
 
 	var hdr object.Object
 	hdr.SetOwnerID(&usr.ID)
-	hdr.SetContainerID(containerID)
+	hdr.SetContainer(containerID)
```
Comment (smallhive) [inline client/container_statistic_test.go:590]: SetContainerID was more clear

File: bearer/bearer.go [hunk 3]
```diff
@@ -261,7 +262,7 @@ func (b *Token) ForUser(id user.ID) {
 //
 // See also ForUser.
 func (b Token) AssertUser(id user.ID) bool {
-	return !b.targetUserSet || b.targetUser == id
+	return b.targetUser.IsZero() || b.targetUser == id
```
Comment (smallhive) [inline bearer/bearer.go:265]: May be change `b.targetUser == id` -> `b.targetUser.Equals(id)`?

File: client/container.go [hunk 4]
```diff
@@ -509,8 +509,7 @@ func (c *Client) ContainerSetEACL(ctx context.Context, table eacl.Table, signer
 		return ErrMissingSigner
 	}
 
-	_, isCIDSet := table.CID()
-	if !isCIDSet {
+	if table.LimitedContainer().IsZero() {
```
Comment (smallhive) [inline client/container.go:512]: `table.CID().IsZero()` is more clear, in my opinion

**Review by smallhive — COMMENTED** [review]
I like we decide to remove fields `isSetSomething`. But I don't like the suggested names for some methods. Previous names more clear and understandable, in my opinion

File: bearer/bearer_test.go (hunk 1 above)
Comment (cthulhu-rider) [reply bearer/bearer_test.go:223]: for me it's not a problem. Usually u state "in a container <id>". But ok i'll change

File: bearer/bearer.go (hunk 3 above)
Comment (cthulhu-rider) [reply bearer/bearer.go:265]: `Equals` deprecated already

File: client/container.go (hunk 4 above)
Comment (cthulhu-rider) [reply client/container.go:512]: this would require interface breakage. I'll leave it `GetCID` then

File: bearer/bearer.go (hunk 3 above)
Comment (smallhive) [reply bearer/bearer.go:265]: Ok, I see

File: client/container.go (hunk 4 above)
Comment (smallhive) [reply client/container.go]: I like `GetCID`

**Review by smallhive — APPROVED** [review]

**Review by carpawell — APPROVED** [review]

File: container/id/id.go [hunk 5]
```diff
@@ -139,3 +148,13 @@ func (id ID) String() string {
 // See also [container.Container.CalculateID], [container.Container.AssertID].
 // Deprecated: use [NewFromContainerBinary].
 func (id *ID) FromBinary(cnr []byte) { *id = NewFromMarshalledContainer(cnr) }
+
+// IsZero checks whether ID is zero.
+func (id ID) IsZero() bool {
+	for i := range id {
```
Comment (roman-khimov) [inline container/id/id.go:154]: `return id == ID{}`?

**Review by roman-khimov — APPROVED** [review]

File: container/id/id.go (hunk 5 above)
Comment (cthulhu-rider) [reply container/id/id.go:154]: no other `ID` needed

Comment (roman-khimov) [reply container/id/id.go:154]: 1. It's less code.
2. It's simpler code.
3. Compiler is likely to detect this and produce a bit better code.
```
$ cat pkg/util/uint256b_test.go
package util

import (
        "testing"
)

func (u Uint256) IsZero() bool {
        for i := range u {
                if u[i] != 0 {
                        return false
                }
        }
        return true
}

func BenchmarkUint256IsZero(b *testing.B) {
        var us = []Uint256{{}, {1}, {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1}}

        for i := 0; i < b.N; i++ {
                for j := range us {
                        _ = us[j].IsZero()
                }
        }
}
```
That's `loop`, and `simple` is `return u == Uint256{}`. Which gives us:
```
$ benchstat iszero.loop iszero.simple 
goos: linux
goarch: amd64
pkg: github.com/nspcc-dev/neo-go/pkg/util
cpu: AMD Ryzen 7 PRO 7840U w/ Radeon 780M Graphics  
                 │ iszero.loop  │            iszero.simple            │
                 │    sec/op    │   sec/op     vs base                │
Uint256IsZero-16   18.610n ± 4%   9.130n ± 1%  -50.94% (p=0.000 n=10)

                 │ iszero.loop │         iszero.simple          │
                 │    B/op     │    B/op     vs base            │
Uint256IsZero-16    0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal

                 │ iszero.loop │         iszero.simple          │
                 │  allocs/op  │ allocs/op   vs base            │
Uint256IsZero-16    0.000 ± 0%   0.000 ± 0%  ~ (p=1.000 n=10) ¹
¹ all samples are equal
```

Comment (cthulhu-rider) [reply container/id/id.go:154]:
```go
func BenchmarkID_IsZero(b *testing.B) {
	var lastByte oid.ID
	lastByte[31]++
	for _, tc := range []struct {
		name string
		id   oid.ID
	}{
		{"zero", oid.ID{}},
		{"first byte", oid.ID{1}},
		{"last byte", lastByte},
	} {
		b.Run(tc.name, func(b *testing.B) {
			b.ReportAllocs()
			for i := 0; i < b.N; i++ {
				tc.id.IsZero()
			}
		})
	}
}
```
```
goos: linux
goarch: amd64
pkg: github.com/nspcc-dev/neofs-sdk-go/object/id
cpu: Intel(R) Core(TM) i5-10210U CPU @ 1.60GHz
                       │    cmp.txt     │               loop.txt                │
                       │     sec/op     │    sec/op      vs base                │
ID_IsZero/zero-8          6.029n ±  17%    8.989n ±  0%  +49.10% (p=0.000 n=10)
ID_IsZero/first_byte-8   3.1970n ±  22%   0.9959n ± 21%  -68.85% (p=0.000 n=10)
ID_IsZero/last_byte-8     5.756n ± 107%    8.759n ±  0%        ~ (p=0.481 n=10)
geomean                   4.805n           4.280n        -10.92%
```

but im more about implementing a method on a type without using the type itself. Now it seems that too much time has passed for such a doit, I don't care what the code will be

**Pull request merged**